etw/logrus/hook.go

```diff
@@ -0,0 +1,78 @@
+package hook
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/Microsoft/go-winio/etw"
+	"github.com/sirupsen/logrus"
+
+	"golang.org/x/sys/windows"
+)
+
+// Hook is a Logrus hook which logs received events to ETW.
+type Hook struct {
+	provider *etw.Provider
+}
+
+// NewHook registers a new ETW provider and returns a hook to log from it.
+func NewHook(providerName string, providerID *windows.GUID) (*Hook, error) {
+	hook := Hook{}
+
+	provider, err := etw.NewProvider(providerName, providerID, nil)
+	if err != nil {
+		return nil, err
+	}
+	hook.provider = provider
+
+	return &hook, nil
+}
+
+// Levels returns the set of levels that this hook wants to receive log entries
+// for.
+func (h *Hook) Levels() []logrus.Level {
+	return []logrus.Level{
+		logrus.TraceLevel,
+		logrus.DebugLevel,
+		logrus.InfoLevel,
+		logrus.WarnLevel,
+		logrus.ErrorLevel,
+		logrus.FatalLevel,
+		logrus.PanicLevel,
+	}
+}
+
+// Fire receives each Logrus entry as it is logged, and logs it to ETW.
+func (h *Hook) Fire(e *logrus.Entry) error {
+	descriptor := etw.NewEventDescriptor()
+
+	// We could try to map Logrus levels to ETW levels, but we would lose some
+	// fidelity as there are fewer ETW levels. So instead we use the level
+	// directly.
+	descriptor.Level = etw.Level(e.Level)
+
+	event := etw.NewEvent("LogrusEntry", descriptor)
+
+	event.Metadata.AddField("Message", etw.InTypeAnsiString)
+	event.Data.AddString(e.Message)
+
+	for k, v := range e.Data {
+		switch reflect.TypeOf(v).Kind() {
+		case reflect.String:

```

Comment:
Can we at least handle all the `InType` in the list before using a generic format string?

Reply:
Also do a type switch rather than this reflection.

Reply:
Will do.